[Docs] add: 8bit inference with pixart alpha

sayakpaul opened the pull request:
# What does this PR do?

This PR adds a section to the PixArt Alpha docs about how to perform inference in 8bit so that the GPU memory requirements are kept under 8GB. 

Cc: @lawrence-cj

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
 
 </Tip>
 
+## Running the `PixArtAlphaPipeline` in under 8GB GPU VRAM
```
Let's put this content into a Tips section (like how it is [here](https://huggingface.co/docs/diffusers/main/en/api/pipelines/audioldm2#tips)) so it is more aligned with the other model API docs. I think this'll improve the docs experience since all the pages will be consistent and users know what to expect if they're going from one page to the next

sayakpaul replied:
I don't think this section is a good candidate to be put as a tip, though. Taking [this](https://huggingface.co/docs/diffusers/main/en/api/pipelines/audioldm2#tips) as an example, I see all the points as advice (or tips). They very well align with what's supposed to be a tip. But I don't think execution under constrained environments can be a tip.

stevhliu replied:
Good point, maybe we can put it in a Usage section then? I think it's better to have something more generalizable to keep the pages consistent (like [DeepFloyd IF](https://huggingface.co/docs/diffusers/main/en/api/pipelines/deepfloyd_if#usage) or [Pix2Pix](https://huggingface.co/docs/diffusers/main/en/api/pipelines/pix2pix_zero#usage-example)).

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
 
+## Running the `PixArtAlphaPipeline` in under 8GB GPU VRAM
+
+It is possible to run the [`PixArtAlphaPipeline`] under 8GB GPU VRAM by loading the text encoder in 8-bit numerical precision. Let's walk through a full-fledged example. 
```
```suggestion
Run the [`PixArtAlphaPipeline`] with under 8GB GPU VRAM by loading the text encoder in 8-bit precision. Let's walk through a full-fledged example. 
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+
+It is possible to run the [`PixArtAlphaPipeline`] under 8GB GPU VRAM by loading the text encoder in 8-bit numerical precision. Let's walk through a full-fledged example. 
+
+First, install the `bitsandbytes` library:
```
```suggestion
First, install the [bitsandbytes](https://github.com/TimDettmers/bitsandbytes) library:
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
````diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+flush()
+```
+
+`flush()` is just a utility function to clear the GPU VRAM and is implemented like so:
````
```suggestion
Free up some GPU VRAM with the `flush()` utility function:
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+    prompt = "cute cat"
+    prompt_embeds, prompt_attention_mask, negative_embeds, negative_prompt_attention_mask = pipe.encode_prompt(prompt)
+
+del text_encoder
+del pipe
+flush()
```
Should this be in the code snippet below where you implement `flush()`? It's being used here without being implemented yet

sayakpaul replied:
I think it's okay this way. Because this way, we immediately make our point clear to the readers and don't introduce any distractions.

sayakpaul replied:
Got it. Reworded.

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
````diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+    torch.cuda.empty_cache()
+```
+
+Then compute the latents providing the prompt embeddings as inputs:
````
```suggestion
Then compute the latents with the prompt embeddings as inputs:
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
````diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+flush()
+```
+
+Notice that while initializing `pipe`, you're setting `text_encoder` to `None` so that it's not loaded. 
````
This could be a `Tip`

```suggestion
<Tip>

Notice that while initializing `pipe`, you're setting `text_encoder` to `None` so that it's not loaded.

</Tip>
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+
+Notice that while initializing `pipe`, you're setting `text_encoder` to `None` so that it's not loaded. 
+
+Once the latents are computed, pass it off the VAE to decode into a real image:
```
```suggestion
Once the latents are computed, pass it off to the VAE to decode into a real image:
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
````diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+image.save("cat.png")
+```
+
+All of this, put together, should allow you to run [`PixArtAlphaPipeline`] under 8GB GPU VRAM.
````
```suggestion
By deleting components you aren't using and flushing the GPU VRAM, you should be able to run [`PixArtAlphaPipeline`] with under 8GB GPU VRAM.
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+
+![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/pixart/8bits_cat.png)
+
+Find the script [here](https://gist.github.com/sayakpaul/3ae0f847001d342af27018a96f467e4e) that can be run end-to-end to report the memory being used.
```
```suggestion
If you want a report of your memory-usage, run this [script](https://gist.github.com/sayakpaul/3ae0f847001d342af27018a96f467e4e).
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+
+<Tip warning={true}>
+
+Text embeddings computed in 8-bit can have an impact on the quality of the generated images because of the information loss in the representation space induced by the reduced precision. It's recommended to compare the outputs with and without 8-bit.
```
```suggestion
Text embeddings computed in 8-bit can impact the quality of the generated images because of the information loss in the representation space caused by the reduced precision. It's recommended to compare the outputs with and without 8-bit.
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,105 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
 
 </Tip>
 
+## Running the `PixArtAlphaPipeline` in under 8GB GPU VRAM
```
```suggestion
## Inference with under 8GB GPU VRAM
```

stevhliu reviewed (approved):
Very cool!

sayakpaul commented:
@stevhliu added a note on 4bit inference in 1265b9de.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_5814). All of your documentation changes will be reflected on that endpoint.

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,111 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+with torch.no_grad():
+    image = pipe.vae.decode(latents / pipe.vae.config.scaling_factor, return_dict=False)[0]
+image = pipe.image_processor.postprocess(image, output_type="pil")
+image.save("cat.png")
```
I'm getting the following error, and I'm unable to save the image:

```py out
AttributeError                            Traceback (most recent call last)
[<ipython-input-7-7a4b72d36ecf>](https://localhost:8080/#) in <cell line: 4>()
      2     image = pipe.vae.decode(latents / pipe.vae.config.scaling_factor, return_dict=False)[0]
      3 image = pipe.image_processor.postprocess(image, output_type="pil")
----> 4 image.save("cat.png")

AttributeError: 'list' object has no attribute 'save'
```

stevhliu commented on docs/source/en/api/pipelines/pixart.md:
```diff
@@ -35,6 +35,111 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
+
+</Tip>
+
+While loading the `text_encoder`, you set `load_in_8bit` to `True`. Instead of `load_in_8bit`, if you specify `load_in_4bit`, the memory requirements will be down under 7GB. 
```
```suggestion
While loading the `text_encoder`, you set `load_in_8bit` to `True`. You could also specify `load_in_4bit` to bring your memory requirements down even further to under 7GB.
```

sayakpaul commented:
@stevhliu I have addressed your comments in https://github.com/huggingface/diffusers/pull/5814/commits/29674b66e1b034582c2405114f9be192152e37bf. I feel good about how the docs are as is right now. Feel free to propose changes in a follow-up PR. 

Going to merge once the CI is green.
